asciidoctorj-core/src/main/java/org/asciidoctor/jruby/cli/AsciidoctorInvoker.java

```diff
@@ -197,10 +197,10 @@ private String readInputFromStdIn() {
 
         List<String> parameters = asciidoctorCliOptions.getParameters();
 
-        if (parameters.isEmpty()) {
-            System.err.println("asciidoctor: FAILED: input file missing");
+        if (parameters.stream().anyMatch(String::isEmpty)) {
```

Comment:
When I invoke `asciidoctorj` without any parameters I get an exception now instead of the error message below:
```
> ./asciidoctorj -r asciidoctor-diagram -b pdf
asciidoctor: FAILED: input file(s) '[]' missing or cannot be read
Exception in thread "main" java.lang.IllegalArgumentException: asciidoctor: FAILED: input file(s) '[]' missing or cannot be read
	at org.asciidoctor.jruby.cli.AsciidoctorInvoker.invoke(AsciidoctorInvoker.java:56)
	at org.asciidoctor.jruby.cli.AsciidoctorInvoker.main(AsciidoctorInvoker.java:229)
```
Can you give some insight why you changed this condition?

Reply:
Before cbeust/jcommander#306 was fixed, the [pre-existing unit test](https://github.com/asciidoctor/asciidoctorj/blob/0d0caee84cd5bce01cbc7d11df85359dc7a0845d/asciidoctorj-core/src/test/java/org/asciidoctor/jruby/cli/WhenAsciidoctorIsCalledUsingCli.java#L126) happens to cover:
* the passing of _empty_ file name
* the passing of _no_ file names

The latter (passing of _no file name_) was actually a side effect of cbeust/jcommander#306 and did not match the actual test implementation (it is passing an **empty** file name).

Once cbeust/jcommander#306 got fixed, the disambiguated _passing of no file names_ case was no longer covered by the unit test.

I'll adjust the PR description / previous commit message (it seems not clear enough that the question is raised) and add a new unit test to covert the _no file names_ use case.

Reply:
@robertpanzer, could you please specify the actual exception you are getting as I am able to get the expected error message you mentioned?

Reply:
@Fiouz My first comment included the exception that I got when calling the asciidoctorj cli. 
It looks like it calls Asciidoctor to convert the files, but passes an empty list on which Asciidoctor Ruby fails. 
But I can rerun and debug once I get my hands on a machine again. 
(Typing on a small screen right now)

Reply:
As far as I understand, the specified command does not reach JRuby, it's only about arguments handling and the "improved" semantic/library.

With current `master` 50b0608c046d3965d65a11d8c06c68d194b79e06
```
$ ./asciidoctorj -r asciidoctor-diagram -b pdf
Picked up JAVA_TOOL_OPTIONS: "-XX:-UsePerfData" "-Duser.language=en" "-Duser.country=US" "-Duser.timezone=UTC" "-Dfile.encoding=UTF-8"
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.headius.backport9.modules.Modules (file:/.../asciidoctorj-distribution/build/distributions/asciidoctorj-2.1.1-SNAPSHOT/lib/jruby-complete-9.2.7.0.jar) to field java.io.FileDescriptor.fd
WARNING: Please consider reporting this to the maintainers of com.headius.backport9.modules.Modules
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
asciidoctor: FAILED: input file missing
Exception in thread "main" java.lang.IllegalArgumentException: asciidoctor: FAILED: input file missing
        at org.asciidoctor.jruby.cli.AsciidoctorInvoker.getInputFiles(AsciidoctorInvoker.java:202)
        at org.asciidoctor.jruby.cli.AsciidoctorInvoker.invoke(AsciidoctorInvoker.java:48)
        at org.asciidoctor.jruby.cli.AsciidoctorInvoker.main(AsciidoctorInvoker.java:229)

$ ./asciidoctorj -r asciidoctor-diagram -b pdf ""
Picked up JAVA_TOOL_OPTIONS: "-XX:-UsePerfData" "-Duser.language=en" "-Duser.country=US" "-Duser.timezone=UTC" "-Dfile.encoding=UTF-8"
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.headius.backport9.modules.Modules (file:/.../asciidoctorj-distribution/build/distributions/asciidoctorj-2.1.1-SNAPSHOT/lib/jruby-complete-9.2.7.0.jar) to field java.io.FileDescriptor.fd
WARNING: Please consider reporting this to the maintainers of com.headius.backport9.modules.Modules
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
asciidoctor: FAILED: input file missing
Exception in thread "main" java.lang.IllegalArgumentException: asciidoctor: FAILED: input file missing
        at org.asciidoctor.jruby.cli.AsciidoctorInvoker.getInputFiles(AsciidoctorInvoker.java:202)
        at org.asciidoctor.jruby.cli.AsciidoctorInvoker.invoke(AsciidoctorInvoker.java:48)
        at org.asciidoctor.jruby.cli.AsciidoctorInvoker.main(AsciidoctorInvoker.java:229)

```

That behavior seems broken:
* it is affected by cbeust/jcommander#306 and never triggers https://github.com/asciidoctor/asciidoctorj/blob/0d0caee84cd5bce01cbc7d11df85359dc7a0845d/asciidoctorj-core/src/main/java/org/asciidoctor/jruby/cli/AsciidoctorInvoker.java#L50-L58 which seems designed to handle the specified command with no file _name_ provided (that piece of code predates this PR)
* it complains that the _file_ is missing whereas it's actually the file _name_ that is missing (checking for the existence of a _file_ without knowing its name is not possible here)
* it complains that the _file_ is missing whereas the issue is on the (bogus) empty file _name_ (a file with an impossible name should not be possible in the first place)

With this PR:
```
$ ./asciidoctorj -r asciidoctor-diagram -b pdf
Picked up JAVA_TOOL_OPTIONS: "-XX:-UsePerfData" "-Duser.language=en" "-Duser.country=US" "-Duser.timezone=UTC" "-Dfile.encoding=UTF-8"
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.headius.backport9.modules.Modules (file:/.../asciidoctorj-distribution/build/distributions/asciidoctorj-2.1.1-SNAPSHOT/lib/jruby-complete-9.2.7.0.jar) to field java.io.FileDescriptor.fd
WARNING: Please consider reporting this to the maintainers of com.headius.backport9.modules.Modules
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
asciidoctor: FAILED: input file(s) '[]' missing or cannot be read
Exception in thread "main" java.lang.IllegalArgumentException: asciidoctor: FAILED: input file(s) '[]' missing or cannot be read
        at org.asciidoctor.jruby.cli.AsciidoctorInvoker.invoke(AsciidoctorInvoker.java:56)
        at org.asciidoctor.jruby.cli.AsciidoctorInvoker.main(AsciidoctorInvoker.java:229)

$ ./asciidoctorj -r asciidoctor-diagram -b pdf ""
Picked up JAVA_TOOL_OPTIONS: "-XX:-UsePerfData" "-Duser.language=en" "-Duser.country=US" "-Duser.timezone=UTC" "-Dfile.encoding=UTF-8"
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.headius.backport9.modules.Modules (file:/C:/.../asciidoctorj-distribution/build/distributions/asciidoctorj-2.1.1-SNAPSHOT/lib/jruby-complete-9.2.7.0.jar) to field java.io.FileDescriptor.fd
WARNING: Please consider reporting this to the maintainers of com.headius.backport9.modules.Modules
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
asciidoctor: FAILED: empty input file name
Exception in thread "main" java.lang.IllegalArgumentException: asciidoctor: FAILED: empty input file name
        at org.asciidoctor.jruby.cli.AsciidoctorInvoker.getInputFiles(AsciidoctorInvoker.java:202)
        at org.asciidoctor.jruby.cli.AsciidoctorInvoker.invoke(AsciidoctorInvoker.java:48)
        at org.asciidoctor.jruby.cli.AsciidoctorInvoker.main(AsciidoctorInvoker.java:229)
```

* it makes use of the previously unreachable https://github.com/asciidoctor/asciidoctorj/blob/0d0caee84cd5bce01cbc7d11df85359dc7a0845d/asciidoctorj-core/src/main/java/org/asciidoctor/jruby/cli/AsciidoctorInvoker.java#L50-L58
* it properly let the user know that the error is about the _name_, not about the absence of file on the filesystem

Reply:
Thanks, now I got it!
I'm fine for this now.
Maybe we can enhance the error message in the future as a further improvement to distinguish the case where no input files were passed as parameters, and where the DirectoryWalker does not match any input file for a given parameter.